### PR TITLE
[MCKIN-7389-js-i18n] Adding i18n support in JS files

### PR DESCRIPTION
**Pull request by afzaledx**
This PR adds support of i18n in Javascript of the DnDv2 xblock. Previously, the translation strings present in text.po and text.mo were not used for strings in JS. WIth the use of namespaced catalog files generated by [django-statici18n](https://github.com/zyegfryed/django-statici18n/pull/41). the translated strings are available by using the function window.DragAndDropI18N.gettext and window.DragAndDropI18N.ngettext

**JIRA tickets**: OSPR ticket

**Screenshots**: 
LMS view:
screen shot 2018-05-17 at 12 49 20 pm

Studio / Editing Views:
screen shot 2018-05-17 at 12 49 54 pm
screen shot 2018-05-17 at 12 50 33 pm
screen shot 2018-05-17 at 12 51 01 pm

**Sandbox URL**: https://studio-pr18208.sandbox.opencraft.hosting/

**Testing instructions**:
1. As staff on a course, add the Drag and Drop problem to a course. It has already been added [here:](https://studio-pr18208.sandbox.opencraft.hosting/container/block-v1:OpenCraft+DnDv2+2018_1+type@vertical+block@ec13f23634824b94a6ddbf4d92afd244)
2. Change the language by visiting the [update_lang](https://studio-pr18208.sandbox.opencraft.hosting/update_lang/) URL. enter 'eo' (without quotes) for language code and click the submit button.
3. Visit the [DnDv2 problem](https://studio-pr18208.sandbox.opencraft.hosting/container/block-v1:OpenCraft+DnDv2+2018_1+type@vertical+block@ec13f23634824b94a6ddbf4d92afd244) page again and refresh the page. All strings should appear translated.
4. Click the Edit button to edit the settings of the DnDv2 problem. All strings should be appear translated.

**Author notes and concerns**:

1. Some strings have not been extracted into the .po/.mo files yet, and so they appear un-translated. They will be marked and translated in another iteration. Exporting missing translation strings is out of the scope of this PR.

2. Any text stored against an XBlocks won't be translated because it's considered "content". These initial feedback strings noted here are provided as suggestions to the educator, and can only be presented exactly as specified by the educator.

**Reviewers**
- [x] (@pomegranited )
- [ ] edX reviewer[s] TBD

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @afzaledx! I've created [OSPR-2405](https://openedx.atlassian.net/browse/OSPR-2405) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



If you like, you can add yourself to the [AUTHORS](https://github.com/edx-solutions/xblock-drag-and-drop-v2/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

File: drag_and_drop_v2/public/js/drag_and_drop_edit.js [hunk 1]
```diff
@@ -1,9 +1,7 @@
 function DragAndDropEditBlock(runtime, element, params) {
 
-    // Set up gettext in case it isn't available in the client runtime:
-    if (typeof gettext == "undefined") {
-        window.gettext = function gettext_stub(string) { return string; };
-    }
+    window.DragAndDropI18n.init();
+    var gettext = window.DragAndDropI18n.gettext;
```
[the remark this answers is not in the document]
Comment (pomegranited) [reply drag_and_drop_v2/public/js/drag_and_drop_edit.js]: Also need to add this to `drag_and_drop.js`, so the translated strings are available to the LMS view.

**Comment by afzaledx** [comment]
@pomegranited, about the fix to the defaults in place of user editable data, I think there might be a solution.

The template uses something like this to display the value of the data:
<textarea class="intro-feedback">{{ self.data.feedback.start }}</textarea>

just replacing the occurances of all titles, descriptions surrounded with the trans tag might solve the issue:
<textarea class="intro-feedback">{% trans self.data.feedback.start %}</textarea>

the defaults will be found in the po/mo and get translated. Any edits made to the text, especially in the foreign language, will surely not find a match and will be displayed untranslated, which is what we want. I have tried it and it seems to be working. I'll be pushing the code shortly.

**Comment by afzaledx** [comment]
@pomegranited , I think I have managed to show almost all of the strings in both the student as well as the studio views.

I could not handle 3 strings:

"Correct! This one belongs to The Top Zone."
"Correct! This one belongs to The Middle Zone."
"Correct! This one belongs to The Bottom Zone."

as they are pre-rendered, and although the template is translated, the rendered string is not found when looked up. Luckily there are just 3 of these and can easily be added to to mo/po files.

Other than that I have tried to address all the situations that you mentioned in your points.

I still have to fix the tests but I've run out of batteries and need to sleep. The changed text.js files (with init() call ) have been included as well.

Feedback is welcome.

**Comment by afzaledx** [comment]
screen shot 2018-05-16 at 1 43 19 am
screen shot 2018-05-16 at 1 43 58 am

File: drag_and_drop_v2/public/js/drag_and_drop.js [hunk 2]
```diff
@@ -1,5 +1,20 @@
...
+        gettext = window.DragAndDropI18N.gettext;
+    } else if ('gettext' in window) {
+        // Use edxapp's global translations
+        gettext = window.gettext;
+    }
+    if (typeof gettext == "undefined") {
+        // No translations -- used by test environment
+        gettext = function gettext_stub(string) { return string; };
+    }
+    window.gettext = gettext;
```
Comment (pomegranited) [inline drag_and_drop_v2/public/js/drag_and_drop.js]: We can't do this -- it overrides the `gettext` for the rest of the javascript rendered on the Open edX page.

**Comment by pomegranited** [comment]
Thanks for fixing those final translation issues @afzaledx :+1:

- [x] I tested this on my devstack using the steps described here: https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/156#pullrequestreview-120048161 
- [x] I read through the code
- [x] I checked for accessibility issues -- a11y is unaffected by this change.
- [x] Includes documentation

File: drag_and_drop_v2/default_data.py [hunk 3]
```diff
@@ -72,7 +74,7 @@
             "displayName": ITEM_TOP_ZONE_NAME,
             "feedback": {
                 "incorrect": ITEM_INCORRECT_FEEDBACK,
-                "correct": ITEM_CORRECT_FEEDBACK.format(zone=TOP_ZONE_TITLE)
```
Comment (afzaledx) [inline drag_and_drop_v2/default_data.py]: @pomegranited , I changed these templated strings to static strings to fix the issue of default data. Hopefully it will not cause problems with translations and don't look that bad either.

Comment (afzaledx) [reply drag_and_drop_v2/default_data.py]: @pomegranited, I have kept this as a separate commit and not extracted the strings into po/mo since I did not know how to do that. perhaps if you could guide me. If it's not required, then I will remove this commit.

Comment (pomegranited) [reply drag_and_drop_v2/default_data.py]: Thanks @afzaledx -- that takes care of the last bit of untranslated text shown on the default demo DnDv2, so this is ok to keep.

Note though that your new strings aren't extracted/compiled into the `.po`/`.mo`/`text.js` files yet, so won't appear translated.  Unfortunately the process for doing this is still very manual, and so this is out of scope here.  We're discussing whether we can update [edx:i18n-tools](https://github.com/edx/i18n-tools) to support managing XBlock translations. CC @bradenmacdonald

**Comment by pomegranited** [comment]
@sanfordstudent This PR is ready for your review CC @mduboseedx

**Comment by sanfordstudent** [comment]
@pomegranited I don't have any expertise in i18n, so would it be possible to get a review from a member of the team that's done internalization work most recently at edX? That's the "Learner <3" team, so I'm tagging one of the team members, @mikix.

**Comment by pomegranited** [comment]
@bradenmacdonald Yep, that's easy enough to do.. we just have to provide a limited list of [settings.LANGUAGES](https://github.com/zyegfryed/django-statici18n/blob/master/src/statici18n/management/commands/compilejsi18n.py#L116) in the settings.py file added here, and remove the `text.js` files for the languages not included in that list.

@afzaledx Could you do this, and include only the languages enabled in [edx-platform's conf/locale/config.yaml](https://github.com/edx/edx-platform/blob/master/conf/locale/config.yaml)?

**Comment by afzaledx** [comment]
@pomegranited @bradenmacdonald , will do.

**Comment by afzaledx** [comment]
@pomegranited @bradenmacdonald , I have reduced the number of languages to 11. There were 10 enabled in the conf/locale/config.yaml file, but I added eo language as well, since it's useful for testing. I will now test to see if it's still working.

**Comment by bradenmacdonald** [comment]
Thanks @afzaledx  and @pomegranited !

**Comment by pomegranited** [comment]
Thanks @afzaledx !  I've retested and your recent change works, and limits the list of languages generated as required.

But I messed up when I removed the `mako` requirement that you added to `requirements.txt`.  When I created a fresh virtualenv to test this, I saw that we need it!  My apologies.  Could you add it back in?

**Comment by mikix** [comment]
Hello, sorry for slow response.

I've done a similar thing in Credentials recently. You can see a series of commits here (it took me a few commits to iron out all the kinks):
https://github.com/edx/credentials/compare/ac4d79ceed15d9555d651fea9933df32835ccb77...ebdc4a048c41727f4d8b81b934733b626ec2ddd1

A few questions for your PR:

1) How do you get translations for the new po files? We use transifex elsewhere in open edx. How would that workflow interact with these files? We might need to create a new Transifex resource for this new po file, outside this PR. (In platform/credentials we use i18n-tools to manage extraction/push/pull)

2) Is there a reason for the preference for text.po rather than the typical djangojs.po? Might help to be more obvious to someone else coming to this code.

3) statici18n does the extraction from JS files for you? That's neat, I didn't know it supported that. Do you happen to know whether it supports es2015/react syntax?

**Comment by bradenmacdonald** [comment]
@mikix  Answering since @pomegranited is away at the moment...

> How do you get translations for the new po files? 

That's something that we'd like to improve - see the discussion near the bottom of [this wiki page from when we discussed this at the recent developer summit](https://openedx.atlassian.net/wiki/spaces/COMM/pages/753532962/Native+XBlocks+Morning+Session). Specifically, it can be done using [this process](https://github.com/edx-solutions/xblock-drag-and-drop-v2#i18n-compatibility) but we really need something better. See also https://github.com/edx/cookiecutter-xblock/pull/4 for compiling the .po files to JS.

> Is there a reason for the preference for text.po rather than the typical djangojs.po? Might help to be more obvious to someone else coming to this code.

That's the standard that was decided on for XBlocks some time ago. I believe it's because XBlocks aren't meant to depend on django, that the "normal" name of text.po was used, instead of a django-specific name. And then I assume it's easier to keep all the translations in that one file instead of splitting them out into JS vs. non-JS translations.

> Do you happen to know whether it supports es2015/react syntax?

No idea, sorry.

**Comment by bradenmacdonald** [comment]
@afzaledx please review what i wrote ^ and add any additional comments you might have.

**Comment by afzaledx** [comment]
@mikix, we had to [modify](https://github.com/zyegfryed/django-statici18n/pull/41) the statici18n to include the namespaced JS required for xblocks. statici18n is just a way to avoid excessive backend calls for translated strings from the JS files and the edx-platform uses that as well.

**Comment by mikix** [comment]
Well I don't have any domain expertise in xblock translations, just some experience with translations at the project scope in open edx. But this PR seems reasonable.

**Comment by pomegranited** [comment]
Thanks for your comments @mikix !  Do we need any further reviews before we can merge this?  CC @mduboseedx

**Comment by pomegranited** [comment]
nudge @mikix  @mduboseedx ^

**Comment by mduboseedx** [comment]
@mikix Could we get a thumbs up for this PR? Jeff also gave it a look over and it looks ok from the i18n perspective

**Comment by edx-webhook** [comment]
Thanks for the pull request, @afzaledx! I've created [OSPR-2532](https://openedx.atlassian.net/browse/OSPR-2532) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



If you like, you can add yourself to the [AUTHORS](https://github.com/edx-solutions/xblock-drag-and-drop-v2/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

**Comment by pomegranited** [comment]
Nudge @mikix CC @mduboseedx

**Comment by pomegranited** [comment]
Thank you for your comments here @mikix !  I'll have a look next week, and ensure that they're addressed.

File: drag_and_drop_v2/public/js/drag_and_drop.js [hunk 4]
```diff
@@ -428,7 +445,7 @@ function DragAndDropTemplates(configuration) {
                 (!ctx.last_action_correct) ? h("p", {}, gettext("Some of your answers were not correct.")) : null,
                 h("p", {}, gettext("Hints:")),
                 h("ul", {}, msgs.map(function(message) {
-                    return h("li", {innerHTML: message.message});
+                    return h("li", {innerHTML: gettext(message.message)});
```
Comment (OmarIthawi) [inline drag_and_drop_v2/public/js/drag_and_drop.js]: @afzaledx I can't see how `makemessages` is supposed to figure out the string to translate?

Comment (afzaledx) [reply drag_and_drop_v2/public/js/drag_and_drop.js]: @OmarIthawi, I am no longer working on this PR. @pomegranited might be able to help you on this.

Comment (OmarIthawi) [reply drag_and_drop_v2/public/js/drag_and_drop.js]: Thanks @afzaledx! Didn't realize that.

Comment (afzaledx) [reply drag_and_drop_v2/public/js/drag_and_drop.js]: No problem. As far as the extraction of strings marked for translation goes, I think it's a manual process so far.

Comment (pomegranited) [reply drag_and_drop_v2/public/js/drag_and_drop.js]: Thanks for replying here @afzaledx 😄  Hope you're doing well.

> I can't see how makemessages is supposed to figure out the string to translate?

@OmarIthawi  This XBlock is kind of a special case, because when you first create a DnDv2 Xblock, we have a demo that displays which uses a bunch of strings stored in [default_data.py](https://github.com/edx-solutions/xblock-drag-and-drop-v2/blob/313fc3b01dd9867687aa430e2b7e5e6da4627f84/drag_and_drop_v2/default_data.py), and that `message.message` string may be one of them.  We do this so that educators using the XBlock for the first time can have a translated experience, which helps them to use the XBlock for their own course.

When the educator then customizes it to suit their own needs for the course, they will likely change all of these strings, and they therefore will be outside the compiled translations library, and so will no longer translated.  We unfortunately can't automatically translate strings we don't know about in advance (and haven't come up with a way for educators to safely provide their own dynamically interpreted translation dictionaries), so that's the best we can do for now.

> As far as the extraction of strings marked for translation goes, I think it's a manual process so far.

Yes, that's correct.  There's a a PR in progress to automate this process using `make` targets: https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/172

**Comment by viadanna** [comment]
I have merged #172 here, use the following sandbox for testing:
https://pr137.sandbox.stage.opencraft.hosting/
https://studio-pr137.sandbox.stage.opencraft.hosting/

File: drag_and_drop_v2/default_data.py [hunk 5]
```diff
@@ -19,7 +19,9 @@
 MIDDLE_ZONE_DESCRIPTION = _("Use this zone to associate an item with the middle layer of the triangle.")
 BOTTOM_ZONE_DESCRIPTION = _("Use this zone to associate an item with the bottom layer of the triangle.")
 
-ITEM_CORRECT_FEEDBACK = _("Correct! This one belongs to {zone}.")
+ITEM_CORRECT_FEEDBACK_TOP = _("Correct! This one belongs to The Top Zone.")
+ITEM_CORRECT_FEEDBACK_MIDDLE = _("Correct! This one belongs to The Middle Zone.")
+ITEM_CORRECT_FEEDBACK_BOTTOM = _("Correct! This one belongs to The Bottom Zone.")
 ITEM_INCORRECT_FEEDBACK = _("No, this item does not belong here. Try again.")
 ITEM_NO_ZONE_FEEDBACK = _("You silly, there are no zones for this one.")
```
Comment (ahmedaljazzar) [inline drag_and_drop_v2/default_data.py]: So this is the file that contains the default values of this XBlock strings @pomegranited mentioned in this comment [comment](https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/172#discussion_r212787389) in PR #172.
Answering the question, it looks weird actually to have see variables surrounded by a translation took, and it wasn't obvious that we're translating default variables here. I liked the idea of having default values and strings, but we might improve this in another PR maybe and implement a better approach for having these values translated.

File: drag_and_drop_v2/drag_and_drop_v2.py [hunk 6]
```diff
@@ -253,6 +254,23 @@ def _learner_raw_score(self):
         correct_count, total_count = self._get_item_stats()
         return correct_count / float(total_count)
 
+    @staticmethod
+    def _get_statici18n_js_url():
```
Comment (ahmedaljazzar) [inline drag_and_drop_v2/drag_and_drop_v2.py]: This method has many nesting depth and cases, it'd be much readable if we flatten it more. [Statement nesting depth](https://help.semmle.com/wiki/display/PYTHON/Statement+nesting+depth)

File: requirements.txt [hunk 7]
```diff
@@ -1,3 +1,5 @@
+mako
 git+https://github.com/edx/xblock-utils.git@v1.1.1#egg=xblock-utils==1.1.1
 git+https://github.com/edx/XBlock.git@xblock-0.5.0#egg=XBlock==0.5.0
+django-statici18n==1.8.0
```
Comment (ahmedaljazzar) [inline requirements.txt]: Let's add `git+https://github.com/edx/i18n-tools.git` as well since it's required by the introduced make targets.

File: Makefile [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+	find . -name '*.pyc' -exec rm -f {} +
+	find . -name '*.pyo' -exec rm -f {} +
+	find . -name '*~' -exec rm -f {} +
+	rm -fr build/
+	rm -fr dist/
+	rm -fr *.egg-info
+
+## Localization targets
+
+extract_translations: ## extract strings to be translated, outputting .po files
```
Comment (ahmedaljazzar) [inline Makefile]: Can we add another make target that extracts the translations and pushes them to Transifex, something like `push_translations`?

File: Makefile [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+	cd $(WORKING_DIR) && DJANGO_SETTINGS_MODULE="$(SETTINGS_FILE)" i18n_tool extract
+	cat $(EXTRACTED_DJANGO) $(EXTRACTED_DJANGOJS) > $(EXTRACTED_TEXT)
+	rm $(EXTRACTED_DJANGO) $(EXTRACTED_DJANGOJS)
+
+compile_translations: ## compile translation files, outputting .mo files for each supported language
```
Comment (ahmedaljazzar) [inline Makefile]: Can we add another make target that pulls the translations from Transifex and compiles them, something like `pull_translations`?

**Comment by ahmedaljazzar** [comment]
@viadanna We're already there; here're some last comments in addition to the inline ones to move this forward:

### Enhancements' comments
- The translations section in README.md is outdated and hard to follow, since we’re building targets to reduce the complexity and to automate the process:
  - Add `transifex-client` to the requirements file since its `tx` command is needed to pull the translations as of [Downloading translations from Transifex](https://github.com/edx-solutions/xblock-drag-and-drop-v2/blob/08ff2efa4538ecd15ae0040ad089af028dde54a7/README.md#downloading-translations-from-transifex) section.
  - No need to describe how XBlock i18n in edX works if the link is provided, this will make working with translations in the repo easier. ([i18n compatibility](https://github.com/edx-solutions/xblock-drag-and-drop-v2/blob/08ff2efa4538ecd15ae0040ad089af028dde54a7/README.md#i18n-compatibility) section)
  - Since we're using `i18n_tools` these commands are no longer needed, right?
    - `~/xblock-drag-and-drop-v2/drag_and_drop_v2$ find . -name "*.py" | xargs xgettext --language=python --add-comments="Translators:"`
    - `~/xblock-drag-and-drop-v2/drag_and_drop_v2$ find . -name "*.js" -o  -path ./public/js/vendor -prune -a -type f | xargs xgettext --language=javascript --from-code=utf-8 --add-comments="Translators:"`
    - `~/xblock-drag-and-drop-v2/drag_and_drop_v2$ msgfmt translations/en/LC_MESSAGES/text.po -o translations/en/LC_MESSAGES/text.mo`
  - In [Building fake "Esperanto" translation](https://github.com/edx-solutions/xblock-drag-and-drop-v2/blob/08ff2efa4538ecd15ae0040ad089af028dde54a7/README.md#building-fake-esperanto-translation) we can reduce the amount of text to point the users to created make target as well.
  - I suggest creating a make target for `tx pull -f --mode=reviewed -l en,ar,es_419,fr,he,hi,ko_KR,pt_BR,ru,zh_CN` command as well and reflect that on Downloading translations from Transifex[](https://github.com/edx-solutions/xblock-drag-and-drop-v2/blob/08ff2efa4538ecd15ae0040ad089af028dde54a7/README.md#downloading-translations-from-transifex) section. Do we need to specify or restrict the pulled languages here given that we did so in `tranlations/config.yaml` file?

### Functional comments
- When I edited some translations in both `.js` and `.py` files and tried to push them to Transifex, I got this error:
```
tx INFO: Pushing resource xblockstest.drag-and-drop-v2
tx INFO: Pushing source file (drag_and_drop_v2/translations/en/LC_MESSAGES/text.po)
tx ERROR: Error received from server: Found empty msgid.
tx ERROR: Could not upload source file. You can use --skip to ignore this error and continue the execution.
tx ERROR: Error received from server: Found empty msgid.
```
- I noticed that we have a non-clean state of translations. That's when I run `extract_translations` I see new added strings and old removed ones, can we check this?

That's it for now, will be installing this in an edx-platform instance and test it there.

File: README.md [hunk 10]
```diff
@@ -552,6 +552,17 @@ $ tx pull -f --mode=reviewed -l en,ar,es_419,fr,he,hi,ko_KR,pt_BR,ru,zh_CN
 
 [transifex-client]: https://docs.transifex.com/client/installing-the-client
 
+Javascript translations
```
[the remark this answers is not in the document]
Comment (viadanna) [reply README.md]: 👍

File: drag_and_drop_v2/default_data.py (hunk 5 above)
Comment (viadanna) [reply drag_and_drop_v2/default_data.py]: 👍

File: Makefile (hunk 8 above)
Comment (viadanna) [reply Makefile]: Sure 👍

File: Makefile (hunk 9 above)
Comment (viadanna) [reply Makefile]: Sure 👍

**Comment by viadanna** [comment]
Hi @ahmedaljazzar 

In terms of the requested changes, I believe the latest commits cover it all.

Just a note on:
> When I edited some translations in both .js and .py files and tried to push them to Transifex, I got this error:
> I noticed that we have a non-clean state of translations. That's when I run extract_translations I see new added strings and old removed ones, can we check this?

I believe these are out of the scope of this work, as both PRs deal with the tooling and not the translations themselves, which seem the cause for both problems. Do you agree?

Oh, I'll update the sandbox to test these.

**Comment by ahmedaljazzar** [comment]
@viadanna Only a couple of comments on the new changes.

> I believe these are out of the scope of this work, as both PRs deal with the tooling and not the translations themselves, which seem the cause for both problems. Do you agree?

I'm not sure, honestly. Since this is a tooling, then we need to make sure it works correctly. Even if Transifex is what throughs the error, I think the problem is from the extractor itself. I also tried to extract the original strings in the repo the push failed due to an error in the plural format. I'm not sure we can move forward while broken functionalities and outdated strings.

And yes, thanks for reflecting the changes on the [cookiecutter-xblock](https://github.com/edx/cookiecutter-xblock/pull/4) 😄

File: requirements.txt [hunk 11]
```diff
@@ -1,3 +1,7 @@
+mako
 git+https://github.com/edx/xblock-utils.git@v1.1.1#egg=xblock-utils==1.1.1
 git+https://github.com/edx/XBlock.git@xblock-0.5.0#egg=XBlock==0.5.0
+git+https://github.com/edx/i18n-tools.git
```
[the remark this answers is not in the document]
Comment (viadanna) [reply requirements.txt]: Messed up the requirements commit, fixing it 👍

**Comment by viadanna** [comment]
Ok, I'll check the extraction process.

File: drag_and_drop_v2/public/js/drag_and_drop_edit.js [hunk 12]
```diff
@@ -326,8 +334,8 @@ function DragAndDropEditBlock(runtime, element, params) {
 
                             // Update zone obj
                             var zoneObj = {
-                                title: oldZone.title || 'Zone ' + num,
-                                description: oldZone.description,
+                                title: gettext(oldZone.title) || gettext('Zone {num}').replace('{num}', num),
```
[the remark this answers is not in the document]
Comment (viadanna) [reply drag_and_drop_v2/public/js/drag_and_drop_edit.js]: I don't know about how other languages would handle this, but I think it doesn't hurt to use `ngettext` here.

**Comment by OmarIthawi** [comment]
@viadanna Just wondering what is this PR waiting for and if I can help.

**Comment by xitij2000** [comment]
@sstack22 This is now ready for review.

**Comment by mduboseedx** [comment]
@sstack22 Do you know of someone that could give a review for a drag and drop change?

**Comment by OmarIthawi** [comment]
@mduboseedx I think this PR has been reviewed very intensively. So perhaps we can change the process a little bit and get it merged. It's been blocked for a while.

I think this would be a great start to experiment with, especially given that this PR has been worked on by more than two different teams, which usually raises quality.

Another PR have already been merged with similar structure (https://github.com/open-craft/xblock-poll/pull/43), and haven't had any issues since so I guess it's not that risky to get this one merged as well.

**Comment by mduboseedx** [comment]
@pomegranited Do you feel that this has had the necessary review? I believe you have the ability to merge changes to the edx-solutions org so if this ok by you then please feel free to merge.

**Comment by pomegranited** [comment]
Thanks for the ping @mduboseedx !  I don't have time to re-test today, but I'll do this tomorrow. CC @xitij2000

**Comment by OmarIthawi** [comment]
@mduboseedx @pomegranited @viadanna thank you'all for your work on this!

**Comment by pomegranited** [comment]
Nevermind @viadanna  -- I ended up rebasing this PR and bumping the version to test it, and so have just pushed the results here.  Thanks to you, @afzaledx @ahmedaljazzar @OmarIthawi @mikix @mduboseedx for all your help and patience here!

Will squash and merge once tests pass.

:+1:

- [x] I tested this using the PR instructions, and also followed the README instructions for re-generating and compiling the translation files.  Updates pushed to 5d72159 .
- [x] I read through the code -- had reviewed it thoroughly during my [previous review round](https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/156#issuecomment-389388169), so here, just ensured that all of the comments raised since then were addressed.
- [x] I checked for accessibility issues by testing this in my devstack using the keyboard only.
- [x] Includes documentation

**Comment by viadanna** [comment]
Oh, thanks a lot, @pomegranited 🎉
